tests/unit/processors/test_files_processor.py

```diff
@@ -0,0 +1,115 @@
+                },
+                "dev": {
+                    "protected": False,
+                    "push_access_levels": [{"access_level_description": "developers"}],
+                    "merge_access_levels": [{"access_level_description": "maintainers"}]
+                },
+                "feature": {
+                    "protected": True,
+                    "push_access_levels": [{"access_level_description": "developers"}],
+                    "merge_access_levels": [{"access_level_description": "developers"}]
+                },
+            }
+        )
+        self.processor = FilesProcessor(self.mock_gitlab, None, True) # Config and strict mode don't affect pre_protection_check
+    
+    def run_tests(self):
+        """
+        Run all tests and print results.
+        """
+        self.test_no_changes_needed_on_protected_branch()
+        self.test_changes_needed_on_protected_branch()
+        self.test_unprotected_branch_should_remain_unprotected()
+        self.test_branch_not_found()
+
+    def assert_equal(self, result, expected, test_name):
+        """
+        Helper method to compare result and expected values.
+        """
+        if result == expected:
+            print(f"{test_name}: PASS")
+        else:
+            print(f"{test_name}: FAIL (Expected: {expected}, Got: {result})")
+
+    def test_no_changes_needed_on_protected_branch(self):
+        """
+        Test that no changes are needed for a correctly protected branch.
+        """
+        result = self.processor.pre_protection_check(
+            "project/group", 
+            "main", 
+            desired_protected_state=True,
+            desired_push_access_level="no one",
+            desired_merge_access_level="maintainers"
+        )
+        self.assert_equal(result, False, "test_no_changes_needed_on_protected_branch")
+
+    def test_changes_needed_on_protected_branch(self):
+        """
+        Test that changes are needed for a protected branch with different access levels.
+        """
+        result = self.processor.pre_protection_check(
+            "project/group", 
+            "feature", 
+            desired_protected_state=True,
+            desired_push_access_level="no one",
+            desired_merge_access_level="maintainers"
+        )
+        self.assert_equal(result, True, "test_changes_needed_on_protected_branch")
+
+    def test_unprotected_branch_should_remain_unprotected(self):
+        """
+        Test that unprotected branch should remain unprotected and no changes are needed.
+        """
+        result = self.processor.pre_protection_check(
+            "project/group", 
+            "dev", 
+            desired_protected_state=False,
+            desired_push_access_level="no one",
+            desired_merge_access_level="maintainers"
+        )
+        self.assert_equal(result, False, "test_unprotected_branch_should_remain_unprotected")
+
+    def test_branch_not_found(self):
```

Comment:
Generally these are a good suite of tests for pre_protection_checks, easy to see what is happening and logic tested

Reply:
Updates look good for Tests @chrishill-1 I'll give it another e2e pass this morning

Reply:
Linters (Black and Types) are failing in the checks below, best to get them fixed this morning whilst I do a review